core/kernel/ree_fs_ta.c

```diff
@@ -229,6 +229,31 @@ static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
 	return res;
 }
 
+static TEE_Result check_ta_type_and_size(const struct shdr *shdr, size_t offs,
+					 size_t ta_size)
+{
+	size_t hdr_sz = 0;
+
+	switch (shdr->img_type) {
+	case SHDR_TA:
+		break;
+	case SHDR_BOOTSTRAP_TA:
+		hdr_sz = sizeof(struct shdr_bootstrap_ta);
+		break;
+	case SHDR_ENCRYPTED_TA:
+		hdr_sz = sizeof(struct shdr_bootstrap_ta) +
+			 sizeof(struct shdr_encrypted_ta);
+		break;
+	default:
+		return TEE_ERROR_SECURITY;
+	}
+
+	if (ta_size != offs + hdr_sz + shdr->img_size)
```

Comment:
we should add the signed heade size also I think (from `SHDR_GET_SIZE(shdr)` or like).

Sometthing like:
```c
static TEE_Result check_ta_type_and_size(const struct shdr *shdr, size_t offs,
					 size_t ta_size, uint8_t *ta)
{
	struct shdr_encrypted_ta img_ehdr = { };
	size_t hdr_sz = SHDR_GET_SIZE(shdr);

	switch (shdr->img_type) {
	case SHDR_TA:
		break;
	case SHDR_BOOTSTRAP_TA:
		hdr_sz += sizeof(struct shdr_bootstrap_ta);
		break;
	case SHDR_ENCRYPTED_TA:
		hdr_sz += sizeof(struct shdr_bootstrap_ta);
		memcpy(&img_ehdr, ta + offs + hdr_sz, sizeof(img_ehdr));
		hdr_sz += SHDR_ENC_GET_SIZE(&img_ehdr);
		break;
	default:
		return TEE_ERROR_SECURITY;
	}

	if (ta_size != offs + hdr_sz + shdr->img_size)
		return TEE_ERROR_SECURITY;

	return TEE_SUCCESS;
}
```
Maybe some ADD_OVERFLOW are missing.

Reply:
The initial `SHDR_GET_SIZE(shdr)` isn't needed, that's already added to offs.

Reply:
It was added only for the subkey headers found in the binary, not for the TA image (bootstrap, encrypted or installed) part were looking into here.